### ISSUE TITLE
feature request: parallel processing of files

### ISSUE THREAD

**Issue by palmtop** [backuppc/backuppc#315]
I use backuppc to backup (and archive) mostly from only one server, but the amount of data to backup is larger (~6T). This is a home environment, but I spent a lot of time to upgrade the network to 10 Gbit, but the backup speeds are still at around ~ 15 MiB/sec. I spent some time looking into the code of Backuppc and it turned out that the bottleneck is that the processing of the files is done in only one thread, and because compression in enabled, one core can not compress faster than this. (This is an older machine, but has 4 cores).  

I propose that on the next major release parallel processing could be implemented, making the backup process faster. It could be done in two levels, first when the files arrive for backup, a new thread could be created for each file and also the compression itself can be made faster with paralell compressing like the one used in pigz.

**Comment by craigbarratt** [comment]
It's not likely that compression is a bottleneck.  First, the main parallelism comes from running multiple backups at the same time.  Yes, each backup involves two rsync processes that are each single threaded, but there can be multiple backups running.  Second, and more importantly, for the rsync Xfer method, in the steady state (ie, if client data hasn't changed) there will be no compression or uncompression (assuming default config settings) done during the backup process.

**Comment by palmtop** [comment]
Last summer I spent a lot of time on migrating my system from v3 to v4 and did a lot of tests on the performance. Before the upgrade one backup of my main server took almost 3 days, but the reason for this was that I used the tar method, after switching to rsync, it reduced to ~4-5 hours. 

If I would be backing up a lot of machines then additional parrallel processing would not be needed, but in my case I have some smaller backups and one big backup, and at the end the big backup is always running alone. 

I did a lot of measurements (unfortunately, now I can not find the exact results, and my test system is already disassembled), the raw conclusion was, that independently of the compression factor, if I enabled compression the speed of storing new files was reduced to around 10-15 Mbit/s and if I switch off compression it went up to around 100 Mbit/s (close to the disk speed limit). When I dug into the details it turned out that zlib (if I remember well), the library used for compression is introducing this bottleneck, as it is using only one thread, and because of the structure of backuppc it is processing only one file at a time, so it can not exceed this speed. I spent a lot of time, looking into the source code, trying to figure out if I could do an easy patch to introduce parallel processing, but backuppc is too complex for me to do this in a realistic time frame. 

I was trying two directions:

1. Introduce some kind of parallelism in the zlib library, this helps always, but according to my searches compression depends on previous parts of the file, so it is not trivial to compress different parts of the files parallel. (pigz has implemented this, but I was not sure if the format it produces is compatible with the original backuppc compression format.)
2. Introduce parallelism on the file processing (start a new thread for processing each file), probably this is easier to implement, but the whole processing should be thought over very carefully, for example how to handle if two files processed parallely are the same. This approach has one disadvantage, if you are processing a very large file, where transfer speed is also longer, the other threads may get out of work.

I believe that the best solution would be to implement both methods, may be together with a file structure and format change, which makes it easier to implement.

This is just an idea, it definitely makes the program much-much more complex, and may be there are only a few use cases, where this brings really overall performance increase.
